kubespawner/objects.py

```diff
@@ -357,6 +366,9 @@ def make_pod(
     # The security settings that you specify for a Pod apply to all Containers
     # in the Pod, but settings on the container level can override them.
     #
+    # We also merge any valid options defined in security_context in
+    # each of those. Options that are not valid are ignored.
+    #
```

Comment:
I don't understand the merging comment, but suggest that the logic is:

1. We bootstrap the pod and container security context:
  - Pod: `supplemental_gids`, `fs_gid`
  - Container: `privileged`, `allow_privilege_escalation`
  - (?) set `uid` and `gid`  wherever they were set before, was it on the pod or container level?
2. We let the explicit pod and container_security_context merge into the bootstrapped ones, giving priority to whatever is in the pod_/container_ security context.

Regarding ignoring, I'm not sure what makes most sense, but I'm afraid about plain ignore as I really dislike silent failures that show up during runtime. I'd prefer hard error if its reliable to not fail because a new API provides something more than before, but otherwise warning seems second best, and ignore as the least attractive option.

Reply:
Does V1PodSecurityContext and V1SecurityContext contain all the entries as I listed before? If so then I suggest we fail hard if something not allowed is provided.

__Pod specific__
- [ ] supplementalGroups
- [ ] fsGroup
- [ ] fsGroupChangePolicy
- [ ] sysctls

__Container specific__
- [ ] privileged
- [ ] allowPrivilegeEscalation
- [ ] capabilities
- [ ] procMount
- [ ] readOnlyRootFilesystem

__Common__
- [ ] runAsUser
- [ ] runAsGroup
- [ ] runAsNonRoot
- [ ] seLinuxOptions
- [ ] seccompProfile
- [ ] windowsOptions

Reply:
- The merging comment is a leftover from my previous attempt at making a unifed `security_context`, disregard it.
- I will swap around my code to let the security context take precedence. I am not super sure about it because I believe docker-stacks use a 100 `gid` which you likely would have to add explicitly.
- For ignoring, the kubernetes-client library should raise a TypeError if you supply something unexpected. I can catch it and raise something else, I am not sure what works best.
- The tests are 99% copy-pasted (but correct hopefully).

Reply:
> I will swap around my code to let the security context take precedence. I am not super sure about it because I believe docker-stacks use a 1000 uid / 100 gid which you likely would have to add explicitly.

Hmmm, I don't understand what you consider very well here. I think in this version of kubespawner the default is to not set uid/gid unless explicitly configured on KubeSpawner. And, if k8s gets a pod created without runAsUser or runAsGroup, I think the default depends on the k8s cluster on how it behaves, which makes it something you want to explicitly set.

With the `uid` configuration as we already have, and the new container_security_context.run_as_user, there will be two separate ways to configure it explicitly, where the latter takes precedence.

The JupyterHub helm chart sets KubeSpawners `uid` / `gid` to 1000 / 100 respectively by default btw.

Reply:
> For ignoring, the kubernetes-client library should raise a TypeError if you supply something unexpected. I can catch it and raise something else, I am not sure what works best.

As long as we raise an error, and know that "something unexpected" won't include the API that was listed in k8s 1.20 specifications at least. I know that the python kubernetes client is quite far behind the release schedule, so it would be good to verify it can actually accept for example `seccompProfile` etc.